src/niri.rs

```diff
@@ -3730,45 +3745,78 @@ impl Niri {
 
     pub fn render_for_screencopy(
         &mut self,
-        backend: &mut Backend,
+        renderer: &mut GlesRenderer,
+        is_winit_backend: bool,
         screencopy: Screencopy,
     ) -> anyhow::Result<()> {
+        let _span = tracy_client::span!("Niri::render_for_screencopy");
+
         let output = screencopy.output().clone();
         ensure!(self.output_state.contains_key(&output), "output is missing");
 
         self.layout.update_render_elements(&output);
 
-        backend
-            .with_primary_renderer(move |renderer| {
-                let elements = self
-                    .render(
-                        renderer,
-                        &output,
-                        screencopy.overlay_cursor(),
-                        RenderTarget::ScreenCapture,
-                    )
-                    .into_iter()
-                    .rev();
-
-                let region_loc = screencopy.region_loc();
-                let elements = elements.map(|element| {
-                    RelocateRenderElement::from_element(
-                        element,
-                        region_loc.upscale(-1),
-                        Relocate::Relative,
-                    )
-                });
+        let elements = self.render(
+            renderer,
+            &output,
+            screencopy.overlay_cursor(),
+            RenderTarget::ScreenCapture,
+        );
 
-                let scale = output.current_scale().fractional_scale().into();
-                let transform = output.current_transform();
-                render_to_shm(renderer, screencopy.buffer(), scale, transform, elements)
-                    .context("error rendering to screencopy shm buffer: {err:?}")?;
+        let damage_tracker = self.screencopy_state.damage_tracker_or_insert(&output);
+        let damages = damage_tracker
+            .output_damage_tracker()
+            .damage_output(1, &elements)
+            .context("output has no active mode")?
+            .0;
 
-                screencopy.submit(false);
+        if screencopy.with_damage() && damages.is_none() {
+            // No damage, wait until there is some when redrawing the output.
+            *damage_tracker.awaiting_copy() = Some(screencopy);
+            return Ok(());
+        }
 
-                Ok(())
-            })
-            .context("primary renderer is missing")?
+        let region_loc = screencopy.region_loc();
+        let elements = elements.into_iter().rev().map(|element| {
+            RelocateRenderElement::from_element(element, region_loc.upscale(-1), Relocate::Relative)
+        });
+
+        let scale = output.current_scale().fractional_scale().into();
+        let transform = if is_winit_backend {
+            Transform::Normal
```

Comment:
Having to do this seems a bit weird to me. Like, didn't `grim` work fine without this on the winit backend? If so then with this change `grim` will no longer work fine on the winit backend.

Also, this sounds like it will corrupt the damage reported to wlr-screencopy. I'm not sure if there's any client that uses damage from screencopy though to test it... Could you search if there's a client like that? Because otherwise we'll have to test damage by comparing numerical values.

Reply:
The reason why I had to take it into account is `wl-recorder` would output a flipped-180 version instead. There might be a transform part that I forgot to add, I will have to dig more into it.

Reply:
This is likely a wf-recorder problem. I check against grim as that should be a more established implementation.